lib/network/internal.nix

```diff
@@ -0,0 +1,220 @@
+{ lib }:
+let
+  inherit (lib)
+    concatStringsSep
+    elemAt
+    fold
+    foldl
+    genList
+    mod
+    range
+    reverseList
+    toInt
+    ;
+
+  inherit (builtins) all any toString;
+in
+rec {
+  common = {
+    /**
+      Given a base and exponent, calculates base raised to the exponent.
+
+      # Example
+
+      ```nix
+      pow 2 3
+      => 8
+      ```
+
+      # Type
+
+      ```
+      pow :: Int -> Int -> Int
+      ```
+
+      # Arguments
+
+      - [base] The base.
+      - [exponent] The exponent.
+
+      # Throws
+
+      - If the exponent is less than 0.
+    */
+    pow =
+      base: exponent:
+      if exponent < 0 then
+        throw "lib.network.pow: Exponent cannot be negative."
+      else if exponent == 0 then
+        1
+      else
+        fold (x: y: y * base) base (range 2 exponent);
+  };
+
+  ipv4 = rec {
+    /**
+      Given a prefix length, returns the equivalent subnet mask.
+
+      # Example
+
+      ```nix
+      _prefixToSubnetMask "24"
+      => "255.255.255.0"
+      ```
+
+      # Type
+
+      ```
+      _prefixToSubnetMask :: String -> String
+
+      # Arguments
+
+      - [prefixLength] A prefix length
+    */
+    _prefixToSubnetMask =
+      prefixLength:
+      let
+        prefixLength' = toInt prefixLength;
+      in
+      _encode ((foldl (x: y: 2 * x + 1) 0 (range 1 prefixLength')) * (common.pow 2 (32 - prefixLength')));
+
+    /**
+      Encodes an integer into a valid IPv4 address.
+
+      # Example
+
+      ```nix
+      _encode 0
+      => "0.0.0.0"
+      _encode 3232235521
+      => "192.168.0.1"
+      ```
+
+      # Type
+
+      ```
+      _encode :: Int -> String
+      ```
+
+      # Arguments
+
+      - [num] A decoded integer representation of an IPv4 address.
+
+      # Throws
+
+      - If the argument is less than zero.
+      - If the argument is greater than or equal to 2^32.
+    */
+    _encode =
+      num:
+      if num < 0 then
```

Comment:
With the `_parseCIDR` approach detailed above, this function won't have to check bounds anymore, since you can internally guarantee it won't be called with invalid arguments.

Reply:
For the time being this is true, but I have plans to add things in the future where having the bounds check will be necessary. Currently, the only part that requires it is subnet mask and I've already removed that per your other suggestion. So I'll remove this for now and we can hash it out in a future PR when subnet mask comes back in.

Sound good?